packages/code-du-travail-nlp/api/sem_search.py

```diff
@@ -0,0 +1,146 @@
+import json
+import tensorflow as tf
+import tensorflow_hub as hub
+import numpy as np
+import tf_sentencepiece
+import unidecode
+
+from typing import List
+
+content_path = "./data/content.json"
+stops_path = "./data/stops.txt"
+
+cache_path = "~/Downloads"
+
+stringvec = List[str]
+
+
+def add_slash(c):
+    c["slug"] = "/{}".format(c["slug"])
+    return c
+
+
+class SemSearch():
+
+    def __init__(self, content_path: str, stops_path: str):
+        print("loading data...")
+        with open(content_path, "r") as f:
+            content = list(filter(
+                lambda row: "text" in row and not "code_du_travail" in row.get('slug'), json.load(f)))
+        content = list(map(add_slash, content))
+
+        with open(stops_path, "r") as f:
+            stops = f.read().splitlines()
+
+        # make a hashtable for performance
+        self.stops = {self.strip_accents(k): "" for k in stops}
+
+        self.titles = [c["title"] for c in content]
+        self.context = [c["text"] for c in content]
+        self.slugs = [c["slug"] for c in content]
+
+        print("Indexing {} documents ...".format(str(len(self.slugs))))
+        self.load_graph()
+        self.build_responses()
+
+    def load_graph(self):
+        g = tf.Graph()
+        with g.as_default():
+            self.q_placeholder = tf.placeholder(tf.string, shape=[None])
+            self.r_placeholder = tf.placeholder(tf.string, shape=[None])
+            self.c_placeholder = tf.placeholder(tf.string, shape=[None])
+            module = hub.Module(
+                "https://tfhub.dev/google/universal-sentence-encoder-multilingual-qa/1")
+            self.question_embeddings = module(
+                dict(input=self.q_placeholder),
+                signature="question_encoder", as_dict=True)
+
+            self.response_embeddings = module(
+                dict(input=self.r_placeholder,
+                     context=self.c_placeholder),
+                signature="response_encoder", as_dict=True)
+            init_op = tf.group(
+                [tf.global_variables_initializer(), tf.tables_initializer()])
+        g.finalize()
+        session = tf.Session(graph=g)
+        session.run(init_op)
+        self.session = session
+
```

Comment:
tu as l'air d'utiliser souvent la combinaison `self.remove_stops(self.strip_accents(x))`, ca vaut peut-etre le coup d'en faire une fonction
```suggestion
    def strip_accents_and_remove_stops(self, string: str):
        # tokens = unidecode.unidecode(string.lower()).split()
        tokens = unicodedata.normalize('NFD', string.lower()).encode('ascii', 'ignore').decode('utf-8').split()
        s = self.stops
        return " ".join(filter(lambda _: _ not in s, tokens))
```

Reply:
Mais du coup on perd l'intérêt du hashtable avec not in?